Deploy More SmartPointers in NativePromise.h

anil-hebbar opened the pull request:
#### 06b78e414b58e034038e812c53ec77d1d27d229b
<pre>
Deploy More SmartPointers in NativePromise.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=287910">https://bugs.webkit.org/show_bug.cgi?id=287910</a>

Reviewed by Ryosuke Niwa.

Addressed the remaining smart pointer static analyzer warnings in NativePromise.h

* Source/WTF/wtf/NativePromise.h:
(WTF::NativePromiseRequest::disconnect):

Canonical link: <a href="https://commits.webkit.org/290819@main">https://commits.webkit.org/290819@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a0c2bd15cacf49f4a73b438684de0820ac513a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96116 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41797 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18860 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70023 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27488 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8321 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82513 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50349 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8095 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40921 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83825 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98003 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89777 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18203 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13374 "Found 1 new test failure: imported/w3c/web-platform-tests/file-system-access/getDirectory.https.any.worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78973 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78316 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78171 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19356 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22657 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/51 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11461 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18210 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23579 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112347 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17946 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32638 "Found 17 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.bytecode-cache, microbenchmarks/memcpy-wasm-medium.js.dfg-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-slow-memory, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-eager, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21404 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19731 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -323,8 +323,8 @@ class NativePromiseRequest :  public CanMakeWeakPtr<NativePromiseRequest> {
-        RefPtr { std::exchange(m_callback, nullptr) }->disconnect();
-    }
+        RefPtr callback = std::exchange(m_callback, nullptr);
+        callback->disconnect();    }
```
} is on a wrong line here.

RupinMittal commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1526,12 +1531,16 @@ class NativePromiseProducer final : public ConvertibleToNativePromise {
     // The Producer may be moved to resolve/reject the completion promise.
     // While we expect m_promise to never be null, it would cause a null dereference in the destructor if the destructor was called after a move.
     RefPtr<PromiseType> m_promise;
+    RefPtr<PromiseType> protectedPromise() const
```
We prefer not to mix the declarations of functions and member variables--rather we declare all the functions first and then the variables. Could we move `protectedPromise()` below `assertIsDead()`?

RupinMittal replied:
Same with the other `protectedPromise()` above. Thanks!

cdumez commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -888,7 +889,7 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
             // Issue the request now if the return value of then()/whenSettled() is not used.
             if (m_thenCallback)
-                m_promise->maybeSettle(m_thenCallback.releaseNonNull(), m_logSiteIdentifier);
+                protectedPromise()->maybeSettle(m_thenCallback.releaseNonNull(), m_logSiteIdentifier);
```
Why cannot we mark `m_promise` as `const` instead?

anil-hebbar replied:
I tried that and for some reason the tests break

cdumez commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -967,14 +968,19 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         RefPtr<ThenCallbackType> m_thenCallback;
         const Logger::LogSiteIdentifier m_logSiteIdentifier;
+
+        Ref<NativePromise> protectedPromise()
```
This should not be in the middle of data members.

cdumez commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1526,14 +1532,20 @@ class NativePromiseProducer final : public ConvertibleToNativePromise {
     const Logger::LogSiteIdentifier m_creationSite; // For logging
     NO_UNIQUE_ADDRESS std::conditional_t<AutoRejectNonVoid, RejectValueT, detail::VoidPlaceholder> m_defaultReject;
+
+    RefPtr<PromiseType> protectedPromise() const
```
This should come before the data members but also, could `m_promise` be `const` instead?

anil-hebbar committed:
Deploy More SmartPointers in NativePromise.h
https://bugs.webkit.org/show_bug.cgi?id=287910

Reviewed by Ryosuke Niwa.

Addressed the remaining smart pointer static analyzer warnings in NativePromise.h

* Source/WTF/wtf/NativePromise.h:
(WTF::NativePromiseRequest::disconnect):

Canonical link: https://commits.webkit.org/290819@main
